### ISSUE TITLE
There should be an easy way to run *exactly* the same checks locally as on github workflow

### ISSUE THREAD

**Issue by filmil** [unicode-org/icu4x#334]
For example, running `cargo clippy` locally won't do, because what we actually run as a workflow is this check:

```
/usr/share/rust/.cargo/bin/cargo clippy --message-format=json --all-targets --all-features -- -D warnings
```

So your `cargo clippy` run may be fine, but if you wait on github to run its version, you can get errors that you missed.

We should have a way to say, preferably in a single command: "run locally *exactly* the same checks that we're running on the test infra", as a way to streamline fixing error checks.  Having to keep mental accounting on exact command lines to use or waiting on github workflow are error prone and lengthy.

**Comment by zbraniecki** [comment]
In particular the "-D warnings" always gets me because those are optional warnings that the clippy team decided not to show by default.

**Comment by filmil** [comment]
> In particular the "-D warnings" always gets me because those are optional warnings that the clippy team decided not to show by default.

FWIW while the strictness of the checks (e.g. whether to use `-D warning` or not) is something we can agree on separately, the more pressing issue is to guarantee that local run settings don't diverge from those in the continuous build.

**Comment by sffc** [comment]
Sounds like we want a top-level *.cargo/config* file with the shortcuts in it?

**Comment by filmil** [comment]
> Sounds like we want a top-level _.cargo/config_ file with the shortcuts in it?

Would it be possible to configure our tools somehow that the default way a developer runs checks would have the same effect as infra, without introducing new commands?

For example, have a set of default flags for clippy so that when I type `cargo clippy`, it does the right thing automatically.

**Comment by zbraniecki** [comment]
Should be. Let me try

**Comment by zbraniecki** [comment]
I took a stab at this today and unfortunately the only way seems to be to plaster `#![deny(clippy::all)]` into our files.

Alternatively, we could put `./.cargo/config.toml` with an alias, but trying to make `clippy` be an alias to `["clippy", "--", "-D", "warnings"]` leads to an infinite loop as expected :)

One way out would be to do some form of `cargo clippy-warn`, the other would be to match CI with default `cargo clippy` - I'm not even sure if the `-- -D warning` is all that useful.
You can see the list of clippy warnings here https://rust-lang.github.io/rust-clippy/master/index.html if you unselec all but `warn`.

I'm all in favor of aligning default `cargo clippy` with CI, and quite frankly, I prefer to have more lenient CI with just default `cargo clippy` there (to spot any significant enough mistakes that clippy community warranted to call errors), while leaving some form fo `cargo clippy-strict` alias to locally check for all optional improvements we can be making.

So, my proposal is:

1) Change CI to *remove* `-D warning`
2) Add `cargo clippy-strict` or `cargo clippy-more` with our custom selection of clippy errors that will evolve over time (for example, warnings about missing `const` is not super useful yet, but may be in the future)

Alternative:
1) Plaster our selection of clippies in `lib.rs` of each component

I prefer the former.
Does that sounds good?

**Comment by zbraniecki** [comment]
For the record, during my pre-release cleanup I got to the point where I was able to run:

```
cargo +nightly clippy -- -W clippy::pedantic -A clippy::must_use_candidate -A clippy::cast_possible_wrap -A clippy::cast_possible_truncation -A clippy::cast_lossless -A clippy::cast_sign_loss -A clippy::missing_errors_doc -A clippy::match-bool -A clippy::unseparated_literal_suffix -A clippy::enum_glob_use -A clippy::match_wildcard_for_single_variants -A clippy::needless_pass_by_value -A clippy::default_trait_access -A clippy::module_name_repetitions -A clippy::match_same_arms -A clippy::doc_markdown -A clippy::map_err_ignore -A clippy::shadow_unrelated -A clippy::option_if_let_else -A clippy::wildcard_imports -A clippy::pub_enum_variant_names -A clippy::non_ascii_literal -A clippy::cast_precision_loss -A clippy::explicit_deref_methods -A clippy::similar_names
```

And get a clean result. Whether we want any of those clippies to be enabled is up to us.

**Comment by sffc** [comment]
I remember asking that we add `-D warning` because it helped find some problems in the past, and without it, the CI build doesn't return an error.  When running on the command line, it doesn't matter as much what the exit code is.

My preference is just to make `cargo clippy-all` run `cargo clippy --all-targets --all-features`.  However, on CI, we should continue using `-D warning`.

**Comment by sffc** [comment]
I'm going to steal this since it goes nicely with #139.
